pkg/cri/resource-manager/control/cri/cri.go

```diff
@@ -0,0 +1,141 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cri
+
+import (
+	"fmt"
+
+	"github.com/intel/cri-resource-manager/pkg/cri/client"
+	"github.com/intel/cri-resource-manager/pkg/cri/resource-manager/cache"
+	"github.com/intel/cri-resource-manager/pkg/cri/resource-manager/control"
+	criapi "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+
+	logger "github.com/intel/cri-resource-manager/pkg/log"
+)
+
+const (
+	// CRIController is the name of this controller.
+	CRIController = cache.CRI
+)
+
+// crictl encapsulated the runtime state of our CRI enforcement/controller.
+type crictl struct {
+	cache  cache.Cache
+	client client.Client
+}
+
+// Make sure crictl implements the control.Controller interface.
+var _ control.Controller = &crictl{}
+
+// Our logger instance.
+var log logger.Logger = logger.NewLogger(CRIController)
+
+// Our CRI controller singleton instance.
+var singleton *crictl
+
+// getCRIController returns our singleton CRI controller instance.
+func getCRIController() control.Controller {
+	if singleton == nil {
+		singleton = &crictl{}
+	}
+	return singleton
+}
+
+// Start initializes the controller for enforcing decisions.
+func (ctl *crictl) Start(cache cache.Cache, client client.Client) error {
+	ctl.cache = cache
+	ctl.client = client
+
+	return nil
+}
+
+// Stop shuts down the controller.
+func (ctl *crictl) Stop() {
+}
+
+// PreCreateHook is the CRI controller pre-create hook.
+func (ctl *crictl) PreCreateHook(c cache.Container) error {
+	log.Debug("running pre-create hook for %s...", c.PrettyName())
+
+	request, ok := c.GetCRIRequest()
+	if !ok {
+		return criError("pre-create hook: no pending CRI request")
+	}
+	create, ok := request.(*criapi.CreateContainerRequest)
+	if !ok {
+		return criError("pre-create hook: pending CRI request of wrong type (%T)", request)
+	}
+
+	create.Config.Command = c.GetCommand()
+	create.Config.Args = c.GetArgs()
+	create.Config.Labels = c.GetLabels()
+	create.Config.Annotations = c.GetAnnotations()
+	create.Config.Envs = c.GetCRIEnvs()
+	create.Config.Mounts = c.GetCRIMounts()
+	create.Config.Devices = c.GetCRIDevices()
+	create.Config.Linux.Resources = c.GetLinuxResources()
+
+	return nil
+}
+
+// PreStartHook is the CRI controller pre-start hook.
+func (ctl *crictl) PreStartHook(c cache.Container) error {
+	return nil
+}
+
+// PostStartHook is the CRI controller post-start hook.
+func (ctl *crictl) PostStartHook(c cache.Container) error {
+	return nil
+}
+
+// PostUpdateHook is the CRI controller post-update hook.
+func (ctl *crictl) PostUpdateHook(c cache.Container) error {
+	var update *criapi.UpdateContainerResourcesRequest
+
+	log.Debug("running post-update hook for %s...", c.PrettyName())
+
+	resources := c.GetLinuxResources()
+	if resources == nil {
+		return nil
+	}
+	request, ok := c.GetCRIRequest()
+	if !ok {
```

Comment:
Just a dumb question: when/why does this happen?

Reply:
It shouldn't happen if nothing is screwed up in request processing. I put the check here because a package boundary is involved here: the request is set (associated with the container) in resource-manager, then modified here in the CRI controller (and finally consumed again in the resource-manager).